library/md5.c

```diff
@@ -302,29 +318,46 @@ void mbedtls_md5_finish( mbedtls_md5_context *ctx, unsigned char output[16] )
     last = ctx->total[0] & 0x3F;
     padn = ( last < 56 ) ? ( 56 - last ) : ( 120 - last );
 
-    mbedtls_md5_update( ctx, md5_padding, padn );
-    mbedtls_md5_update( ctx, msglen, 8 );
+    if( ( ret = mbedtls_md5_update_ext( ctx, md5_padding, padn ) ) != 0 )
+            return( ret );
+
+    if( ( ret = mbedtls_md5_update_ext( ctx, msglen, 8 ) ) != 0 )
+            return( ret );
 
     PUT_UINT32_LE( ctx->state[0], output,  0 );
     PUT_UINT32_LE( ctx->state[1], output,  4 );
     PUT_UINT32_LE( ctx->state[2], output,  8 );
     PUT_UINT32_LE( ctx->state[3], output, 12 );
+
+    return( 0 );
 }
 
 #endif /* !MBEDTLS_MD5_ALT */
 
 /*
  * output = MD5( input buffer )
  */
-void mbedtls_md5( const unsigned char *input, size_t ilen, unsigned char output[16] )
+int mbedtls_md5_ext( const unsigned char *input,
+                     size_t ilen,
+                     unsigned char output[16] )
 {
+    int ret;
     mbedtls_md5_context ctx;
 
     mbedtls_md5_init( &ctx );
-    mbedtls_md5_starts( &ctx );
-    mbedtls_md5_update( &ctx, input, ilen );
-    mbedtls_md5_finish( &ctx, output );
+
+    if( ( ret = mbedtls_md5_starts_ext( &ctx ) ) != 0 )
+            return( ret );
+
+    if( ( ret = mbedtls_md5_update_ext( &ctx, input, ilen ) ) != 0 )
+            return( ret );
+
+    if( ( ret = mbedtls_md5_finish_ext( &ctx, output ) ) != 0 )
+            return( ret );
+
     mbedtls_md5_free( &ctx );
+
+    return( 0 );
```

Comment:
@hanno-arm I have a PR in progress to mass-edit the descriptions of `mbedtls_XXX_init`, `mbedtls_XXX_setup`, etc. to clarify this sort of issue. My current wording for `free` functions is:

> This function wipes any sensitive data from the context
> and releases any resource used by the context other than
> the memory used by the context structure itself.
> This function must be called when the context is no
> longer in use, before freeing the context structure or
> letting it run out of scope if the context structure is
> on the stack.

I haven't yet gone through `starts`/`update`/`finish`. I think it should always be safe to call `free` on a context structure, and I should tweak the wording above to make this explicit. Thoughts?